src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts

```diff
@@ -0,0 +1,1048 @@
+        { copyWidthInBlocks: 5, copyHeightInBlocks: 4, copyDepth: 1 }, // copyDepth = 1
+        { copyWidthInBlocks: 7, copyHeightInBlocks: 1, copyDepth: 1 }, // copyHeight = 1 and copyDepth = 1
+      ])
+      .combine(poptions('format', kSizedTextureFormats))
+      .filter(formatCanBeTested)
+  )
+  .fn(async t => {
+    const {
+      bytesPerRowPadding,
+      rowsPerImagePaddingInBlocks,
+      copyWidthInBlocks,
+      copyHeightInBlocks,
+      copyDepth,
+      format,
+      initMethod,
+      checkMethod,
+    } = t.params;
+
+    const info = kSizedTextureFormatInfo[format];
+
+    /** For CopyB2T and CopyT2B we need to have bytesPerRow 256-aligned,
+     * to make this happen we align the bytesInACompleteRow value and multiply
+     * bytesPerRowPadding by 256. */
+    const bytesPerRowAlignment =
+      initMethod === 'WriteTexture' && checkMethod === 'FullCopyT2B' ? 1 : 256;
+
+    const copyWidth = copyWidthInBlocks * info.blockWidth;
+    const copyHeight = copyHeightInBlocks * info.blockHeight;
+    const rowsPerImage = copyHeight + rowsPerImagePaddingInBlocks * info.blockHeight;
+    const bytesPerRow =
+      align(t.bytesInACompleteRow(copyWidth, format), bytesPerRowAlignment) +
+      bytesPerRowPadding * bytesPerRowAlignment;
+    const copySize = { width: copyWidth, height: copyHeight, depth: copyDepth };
+
+    const minDataSize = t.requiredBytesInCopy(
+      { offset: 0, bytesPerRow, rowsPerImage },
+      format,
+      copySize
+    );
+
+    t.uploadTextureAndVerifyCopy({
+      textureDataLayout: { offset: 0, bytesPerRow, rowsPerImage },
+      copySize,
+      dataSize: minDataSize,
+      textureSize: [
+        Math.max(copyWidth, info.blockWidth),
+        Math.max(copyHeight, info.blockHeight),
+        Math.max(copyDepth, 1),
+      ] /** making sure the texture is non-empty */,
+      format,
+      initMethod,
+      checkMethod,
+    });
+  });
+
+/**  Test that copying data with various offset values and additional data paddings
+ * works for every format with 2d and 2d-array textures.
+ * Covers two special code paths for D3D12:
+ *   offset + bytesInCopyExtentPerRow { ==, > } bytesPerRow
+ *  offset > bytesInACompleteCopyImage  */
+g.test('copy_with_various_offsets_and_data_sizes')
+  .params(
+    params()
+      .combine(kMethodsToTest)
+      .combine([
+        { offsetInBlocks: 0, dataPaddingInBytes: 0 } /** no offset and no padding */,
+        { offsetInBlocks: 1, dataPaddingInBytes: 0 } /** offset = 1 */,
+        { offsetInBlocks: 2, dataPaddingInBytes: 0 } /** offset = 2 */,
+        { offsetInBlocks: 15, dataPaddingInBytes: 0 } /** offset = 15 */,
+        { offsetInBlocks: 16, dataPaddingInBytes: 0 } /** offset = 16 */,
+        {
+          offsetInBlocks: 242,
+          dataPaddingInBytes: 0,
+        } /** for rgba8unorm format: offset + bytesInCopyExtentPerRow = 242 + 12 = 256 = bytesPerRow */,
```

Comment:
nit: The jsdoc style (even though it's ignored) implies that the comment applies to the thing after it, which isn't the case here